Camera2SlowMotion/app/src/main/java/com/example/android/camera2/slowmo/ServerHandler.java

```diff
@@ -1,125 +1,110 @@
 package com.example.android.camera2.slowmo;
 
 import android.content.ContentValues;
-import android.os.AsyncTask;
+import android.os.Build.VERSION_CODES;
 import android.util.Log;
+import androidx.annotation.RequiresApi;
 import java.io.BufferedReader;
 import java.io.IOException;
 import java.io.InputStreamReader;
 import java.io.PrintWriter;
+import java.io.UncheckedIOException;
 import java.net.Socket;
 import java.util.ArrayList;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
 import java.util.concurrent.ExecutionException;
-import org.jetbrains.annotations.NotNull;
 
 /** Handles every interaction of app with the server. */
 public class ServerHandler {
-
-  private Integer serverSocketPort;
-  private long serverStartTimestamp;
-  private Socket socket;
-  private BufferedReader inputReader;
-  private PrintWriter outputWriter;
-  private long hostSyncTimestamp;
-
-  public ServerHandler(Integer serverSocketPort) {
-    this.serverSocketPort = serverSocketPort;
+  private static class Connection {
+    Socket socket;
+    BufferedReader inputReader;
+    PrintWriter outputWriter;
   }
 
-  public long getHostSyncTimestamp() {
-    return hostSyncTimestamp;
+  public static class ServerData{
+    ArrayList<Long> timestamps;
+    ArrayList<String> serverSequence;
+    ServerData(){
+      this.timestamps = new ArrayList<>();
+      this.serverSequence = new ArrayList<>();
+    }
   }
 
-  public Integer getServerSocketPort() {
-    return serverSocketPort;
-  }
+  private Integer serverSocketPort;
+  private CompletableFuture<Connection> connection;
 
-  public void setServerSocketPort(Integer serverSocketPort) {
+  public ServerHandler(Integer serverSocketPort) {
     this.serverSocketPort = serverSocketPort;
   }
 
-  public long getServerStartTimestamp() {
-    return serverStartTimestamp;
-  }
-
+  /**
+   * Creates a socket for communication with the server. Sets reader and writer for server
+   * connection.
+   */
+  @RequiresApi(api = VERSION_CODES.N)
   public void startConnection() {
-    ConnectionThread connectionThread = new ConnectionThread();
-    connectionThread.execute();
+    connection =
```

Comment:
I can't see where startConnection() is used from the changes here, but would it be better to to explicitly wait until the connection was established before we started capture? That way, we can handle any errors and don't start sending keypresses late (if the connection took a while to establish).